### PR TITLE
Ret 2256 - create a controller for document upload service

### PR DESCRIPTION
**Pull request by DarylBoam**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RET-2256

### Change description ###
create a controller for document upload service

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/et/syaapi/controllers/DocumentUploadController.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    ) {
+        if (!validateRequest(caseTypeId, multipartFile)) {
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+        }
+        try {
+            return ResponseEntity.ok(caseDocumentService.uploadDocument(authorization, caseTypeId,
+                multipartFile));
+        } catch (CaseDocumentException ex) {
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/et/syaapi/controllers/DocumentUploadController.java:54]: Worth adding a log info/warn here so we capture the exception thrown for future debugging

File: src/main/java/uk/gov/hmcts/reform/et/syaapi/models/CaseDocument.java [hunk 2]
```diff
@@ -14,9 +15,32 @@
...
+    @JsonProperty("ttl")
+    String ttl;
+    @JsonProperty("metadata")
+    Map<String, String> metadata;
     @JsonProperty("_links")
     Map<String, Map<String, String>> links;
+
+    public URI getUri() {
+        return URI.create(links.get("self").get("href"));
+    }
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/et/syaapi/models/CaseDocument.java:45]: This should have JavaDoc as until I read the actual code I didn't know it was getting the URI for the links property named "self"

File: src/main/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentService.java [hunk 3]
```diff
@@ -164,17 +153,21 @@ private void validateFile(MultipartFile file) throws CaseDocumentException, IOEx
...
         }
 
         CaseDocument document = response.getDocuments().stream()
             .findFirst()
             .orElseThrow(() -> new CaseDocumentException(UPLOAD_FILE_EXCEPTION_MESSAGE + originalFilename));
 
-        String uri = getUriFromFile(document, originalFilename).toString();
+        if (document.getLinks() == null
+            || document.getLinks().get("self") == null
+            || document.getUri() == null) {
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentService.java:166]: Validating the "self" link's URL might be better done alongside the getUri method on CaseDocument.  Still verify and throw the exception here, but would add value doing the checks in there

File: src/test/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentServiceTest.java [hunk 4]
```diff
@@ -93,14 +94,21 @@ class CaseDocumentServiceTest {
...
+    private final String fullJsonResponse = ResourceUtil.resourceAsString(
+        "responses/caseDocumentUpload.json"
+    );
 
     private CaseDocumentService caseDocumentService;
     private MockRestServiceServer mockServer;
 
+    CaseDocumentServiceTest() throws IOException {
+        // constructor for case document series test
+    }
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentServiceTest.java:106]: Surprises me you need this???

**Review by nickwest1 — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentServiceTest.java (hunk 4 above)
Comment (DarylBoam) [reply src/test/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentServiceTest.java:106]: It is required for  ResourceUtil.resourceAsString() statement

**Review by robin-ieong — DISMISSED** [review]

**Review by nickwest1 — APPROVED** [review]

**Review by marcjudge — APPROVED** [review]

**Review by ikirsanov — APPROVED** [review]

**Pull request merged**